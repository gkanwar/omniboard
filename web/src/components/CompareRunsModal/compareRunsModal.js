import PropTypes from 'prop-types';
import React, {PureComponent} from 'react';
import {Modal, ModalHeader, ModalBody, ModalTitle, Alert, Nav, NavItem, Well} from 'react-bootstrap';
import './compareRunsModal.scss';
import axios from 'axios';
import Multiselect from 'react-bootstrap-multiselect';
import {ProgressWrapper} from '../Helpers/hoc';
import {parseServerError} from '../Helpers/utils';
import {DRILLDOWN_VIEW} from '../../appConstants/drillDownView.constants';
import {MetricsPlotView} from '../MetricsPlotView/metricsPlotView';
import {CapturedOutCompareView} from '../CapturedOutCompareView/capturedOutCompareView';
import {ConfigCompareView} from '../ConfigCompareView/configCompareView';
import {SourceFilesCompareView} from '../SourceFilesCompareView/sourceFilesCompareView';

class CompareRunsModal extends PureComponent {
  static propTypes = {
    handleClose: PropTypes.func.isRequired,
    shouldShow: PropTypes.bool.isRequired,
    runs: PropTypes.arrayOf(PropTypes.number).isRequired
  };

  runIdOptionsDomNode = null;

  constructor(props) {
    super(props);
    this.state = {
      isLoadingRuns: false,
      selectedNavTab: DRILLDOWN_VIEW.METRICS,
      metrics: [],
      runIdOptions: []
    };
  }

  _fetchMetrics = () => {
    const {runs} = this.props;
    if (runs.length >= 2) {
      this.setState({
        isLoadingRuns: true,
        error: ''
      });
      const queryString = JSON.stringify({
        run_id: {
          $in: runs
        }
      });
      axios.get('/api/v1/Metrics', {
        params: {
          query: queryString
        }
      }).then(response => {
        this.setState({
          metrics: response.data,
          isLoadingRuns: false
        });
      }).catch(error => {
        const message = parseServerError(error);
        this.setState({
          isLoadingRuns: false,
          error: message
        });
      });
    } else {
      this.setState({
        error: 'At-least two runs should be selected for comparison'
      });
    }
  };

  _populateRunIdOptions = () => {
    const {runs} = this.props;
    const runIdOptions = runs.map(runId => {
      return {
        label: runId,
        value: runId,
        selected: true
      };
    });
    this.setState({
      runIdOptions
    });
  };

  componentDidMount() {
    // Load data only when the component is being displayed
    if (this.props.shouldShow === true) {
      this._fetchMetrics();
      this._populateRunIdOptions();
    }
  }

  componentDidUpdate(prevProps, _prevState, _snapshot) {
    // Load data every time this modal is being popped up
    if (prevProps.shouldShow !== this.props.shouldShow && this.props.shouldShow === true) {
      this._fetchMetrics();
      this._populateRunIdOptions();
    }
  }

  _handleSelectNavPill = selectedKey => {
    this.setState({
      selectedNavTab: selectedKey
    });
  };

  _getSelectedRunIds = runIdOptions => runIdOptions.filter(option => option.selected === true)
    .map(option => Number(option.value));

  _handleRunIdsChange = _e => {
    const runIds = this.runIdOptionsDomNode.$multiselect.val();
    this.setState(({runIdOptions}) => ({
      runIdOptions: runIdOptions.map(option => {
        option.selected = runIds.includes(String(option.value));
        return option;
      })
    }));
  };

  render() {
    const {shouldShow, handleClose, runs} = this.props;
    const {isLoadingRuns, error, selectedNavTab, metrics, runIdOptions} = this.state;
    const errorAlert = error ? <Alert bsStyle='danger'>{error}</Alert> : '';
    // When there are more than 5 runs selected, then the title should show up as [1, 2, 3, 4, 5, ...]
    const runIds = runs.length <= 5 ? runs.join(', ') : runs.slice(0, 5).join(', ') + ', ...';
    const modalTitle = `Compare ${runs.length} Runs [${runIds}]`;
    const selectedRunIds = this._getSelectedRunIds(runIdOptions);
    const metricsResponseForPlot = metrics.filter(metric => selectedRunIds.includes(metric.run_id));
    const getTabContent = key => {
      let content = '';
      switch (key) {
        case DRILLDOWN_VIEW.CAPTURED_OUT:
          content = <div id={DRILLDOWN_VIEW.CAPTURED_OUT}><CapturedOutCompareView runIds={runs}/></div>;
          break;
        case DRILLDOWN_VIEW.METRICS:
          content = (
            <div id={DRILLDOWN_VIEW.METRICS}>
              <Well bsSize='small' className='run-id-filter-well'>
                <h5>Runs</h5>
                <div className='run-id-options-wrapper'>
                  <Multiselect
                    ref={el => this.runIdOptionsDomNode = el}
                    includeSelectAllOption multiple
                    enableHTML
                    buttonText={(options, _select) => {
                      if (options.length === 0) {
                        return 'None selected';
                      }

                      return `${options.length} selected`;
                    }}
                    selectedClass='run-id-selected'
                    id='run_ids'
                    maxHeight={300}
                    data={runIdOptions}
                    onSelectAll={this._handleRunIdsChange}
                    onChange={this._handleRunIdsChange}
                    onDeselectAll={this._handleRunIdsChange}
                  />
                </div>
              </Well>
              <MetricsPlotView metricsResponse={metricsResponseForPlot} runId=''/>
            </div>
          );
          break;
        case DRILLDOWN_VIEW.CONFIG:
          content = (
            <div id={DRILLDOWN_VIEW.CONFIG}>
              <ConfigCompareView runIds={runs}/>
            </div>
          );
          break;
        case DRILLDOWN_VIEW.SOURCE_FILES:
          content = (
            <div id={DRILLDOWN_VIEW.SOURCE_FILES}>
              <SourceFilesCompareView runIds={runs} isSelected={selectedNavTab === DRILLDOWN_VIEW.SOURCE_FILES}/>
            </div>
          );
          break;
        default:
      }

      return <ProgressWrapper id='ddv-progress-wrapper' loading={isLoadingRuns}>{content}</ProgressWrapper>;
    };

    return (
      <Modal dialogClassName='compare-runs-modal' show={shouldShow} onHide={handleClose}>
        <ModalHeader closeButton>
          <ModalTitle>{modalTitle}</ModalTitle>
        </ModalHeader>
        <ModalBody>
          {errorAlert}
          <ProgressWrapper loading={isLoadingRuns}>
            <div className='full-height'>
              <div className='sidebar-wrapper full-height'>
                <div className='sidebar-container full-height'>
                  <div id='sidebar'>
                    <Nav stacked bsStyle='pills' activeKey={selectedNavTab} onSelect={this._handleSelectNavPill}>
                      <NavItem eventKey={DRILLDOWN_VIEW.METRICS}>
                        Metrics Plot
                      </NavItem>
                      <NavItem eventKey={DRILLDOWN_VIEW.CAPTURED_OUT}>
                        Captured Out
                      </NavItem>
                      <NavItem eventKey={DRILLDOWN_VIEW.SOURCE_FILES}>
                        Source Files
                      </NavItem>
                      <NavItem eventKey={DRILLDOWN_VIEW.CONFIG}>
                        Config
                      </NavItem>
                    </Nav>
                  </div>
                </div>
                <div className='sidebar-content full-height'>
                  <div ref={div => (this.scrollableDiv = div || this.scrollableDiv)} className='tab-content'>
                    {getTabContent(selectedNavTab)}
                  </div>
                </div>
              </div>
            </div>
          </ProgressWrapper>
        </ModalBody>
      </Modal>
    );
  }
}

export {CompareRunsModal};
