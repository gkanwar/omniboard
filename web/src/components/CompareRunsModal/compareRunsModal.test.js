import React from 'react';
import mockAxios from 'jest-mock-axios';
import {DRILLDOWN_VIEW} from '../../appConstants/drillDownView.constants';
import {CompareRunsModal} from './compareRunsModal';

describe('CompareRunsModal', () => {
  let wrapper = null;
  const responseData = [
    {_id: '5a2b5a8c9c7a505a652f6127', name: 'pretrain.train.loss', run_id: 1, steps: [0, 1, 2, 3, 4],
      values: [0.7159541824544438, 0.3840367944955761, 0.3469185283233073, 0.30483262065173106, 0.28915774130337507],
      timestamps: ['2017-12-09T03:37:44.425Z', '2017-12-09T03:41:54.414Z', '2017-12-09T03:46:01.766Z', '2017-12-09T03:50:07.365Z', '2017-12-09T03:54:12.560Z']},
    {_id: '5a2b5aa09c7a505a652f6146', name: 'pretrain.val.loss', run_id: 2, steps: [0, 1, 2, 3, 4],
      values: [0.32177006650114165, 0.23237958704995795, 0.23340759051386187, 0.21925230575196739, 0.20541178824900605],
      timestamps: ['2017-12-09T03:38:01.945Z', '2017-12-09T03:42:11.673Z', '2017-12-09T03:46:18.843Z', '2017-12-09T03:50:24.377Z', '2017-12-09T03:54:29.752Z']},
    {_id: '5a2b5a8c9c7a505a652f6127', name: 'pretrain.train.loss', run_id: 3, steps: [0, 1, 2, 3, 4],
      values: [0.7859541824544438, 0.5640367944955761, 0.8769185283233073, 0.60483262065173106, 0.58915774130337507],
      timestamps: ['2017-12-09T03:37:44.425Z', '2017-12-09T03:41:54.414Z', '2017-12-09T03:46:01.766Z', '2017-12-09T03:50:07.365Z', '2017-12-09T03:54:12.560Z']},
    {_id: '5a2b5aa09c7a505a652f6146', name: 'pretrain.val.loss', run_id: 4, steps: [0, 1, 2, 3, 4],
      values: [0.52177006650114165, 0.9237958704995795, 0.11340759051386187, 0.41925230575196739, 0.90541178824900605],
      timestamps: ['2017-12-09T03:38:01.945Z', '2017-12-09T03:42:11.673Z', '2017-12-09T03:46:18.843Z', '2017-12-09T03:50:24.377Z', '2017-12-09T03:54:29.752Z']}
  ];
  const runIds = [1, 2, 3, 4];
  const handleCloseMock = jest.fn();
  console.error = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <CompareRunsModal shouldShow runs={runIds} handleClose={handleCloseMock}/>
    );
  });

  afterEach(() => {
    // Cleaning up the mess left behind the previous test
    mockAxios.reset();
    jest.clearAllMocks();
    wrapper = null;
  });

  describe('should load data', () => {
    it('success', async () => {
      expect(mockAxios.get.mock.calls).toHaveLength(1);

      expect(mockAxios.get.mock.calls[0]).toEqual(['/api/v1/Metrics', {params: {query: JSON.stringify({run_id: {$in: [1, 2, 3, 4]}})}}]);
      expect(wrapper.instance().state.isLoadingRuns).toBeTruthy();
      mockAxios.mockResponse({status: 200, data: responseData});

      await tick();
      expect(wrapper.update().instance().state.isLoadingRuns).toBeFalsy();
      expect(wrapper.update()).toMatchSnapshot();
    });

    it('error', () => {
      const errResponse = {status: 400, message: 'Not Found'};
      mockAxios.mockError(errResponse);

      expect(wrapper.instance().state.error).not.toEqual('');
    });
  });

  it('should not load data when runIds < 2', async () => {
    wrapper.unmount();
    mockAxios.reset();
    wrapper = mount(
      <CompareRunsModal shouldShow runs={[1]} handleClose={handleCloseMock}/>
    );

    expect(mockAxios.get.mock.calls).toHaveLength(0);
    expect(wrapper.instance().state.error).not.toEqual('');
  });

  it('should handle run id change', async () => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.reset();
    wrapper.instance().runIdOptionsDomNode = {
      $multiselect: {
        val: () => ['4', '3', '1']
      }
    };
    wrapper.instance()._handleRunIdsChange();

    expect(wrapper.instance().state.runIdOptions.find(option => option.value === 2).selected).toBeFalsy();
  });

  describe('componentDidUpdate', () => {
    beforeEach(() => {
      mockAxios.mockResponse({status: 200, data: responseData});
      mockAxios.reset();
    });

    it('loads data', async () => {
      wrapper.setProps({shouldShow: false});
      wrapper.setProps({shouldShow: true});

      expect(mockAxios.get.mock.calls).toHaveLength(1);
      expect(wrapper.instance().state.isLoadingRuns).toBeTruthy();
    });

    it('does not load data', async () => {
      wrapper.setProps({shouldShow: true});

      expect(mockAxios.get.mock.calls).toHaveLength(0);
    });
  });

  it('should switch tabs correctly', async () => {
    wrapper.unmount();
    mockAxios.reset();
    jest.clearAllMocks();
    wrapper = mount(
      <CompareRunsModal shouldShow runs={runIds} handleClose={handleCloseMock}/>
    );
    mockAxios.mockResponse({status: 200, data: responseData});
    await tick();
    const switchTabTest = tabKey => {
      wrapper.update().instance()._handleSelectNavPill(tabKey);

      expect(wrapper.update().find('div.tab-content ProgressWrapper#ddv-progress-wrapper').instance().props.children.props.id).toEqual(tabKey);
    };

    [
      DRILLDOWN_VIEW.CAPTURED_OUT,
      DRILLDOWN_VIEW.METRICS,
      DRILLDOWN_VIEW.CONFIG,
      DRILLDOWN_VIEW.SOURCE_FILES
    ].forEach(key => switchTabTest(key));
  });
});
