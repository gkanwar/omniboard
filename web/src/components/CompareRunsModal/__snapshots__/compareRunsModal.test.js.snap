// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompareRunsModal should load data success 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogClassName="compare-runs-modal"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[MockFunction]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Compare 4 Runs [1, 2, 3, 4]
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <ProgressWrapper
      loading={false}
    >
      <div
        className="full-height"
      >
        <div
          className="sidebar-wrapper full-height"
        >
          <div
            className="sidebar-container full-height"
          >
            <div
              id="sidebar"
            >
              <Nav
                activeKey="metrics"
                bsClass="nav"
                bsStyle="pills"
                justified={false}
                onSelect={[Function]}
                pullLeft={false}
                pullRight={false}
                stacked={true}
              >
                <NavItem
                  active={false}
                  disabled={false}
                  eventKey="metrics"
                >
                  Metrics Plot
                </NavItem>
                <NavItem
                  active={false}
                  disabled={false}
                  eventKey="captured_out"
                >
                  Captured Out
                </NavItem>
                <NavItem
                  active={false}
                  disabled={false}
                  eventKey="source_files"
                >
                  Source Files
                </NavItem>
                <NavItem
                  active={false}
                  disabled={false}
                  eventKey="config"
                >
                  Config
                </NavItem>
              </Nav>
            </div>
          </div>
          <div
            className="sidebar-content full-height"
          >
            <div
              className="tab-content"
            >
              <ProgressWrapper
                id="ddv-progress-wrapper"
                loading={false}
              >
                <div
                  id="metrics"
                >
                  <Well
                    bsClass="well"
                    bsSize="small"
                    className="run-id-filter-well"
                  >
                    <h5>
                      Runs
                    </h5>
                    <div
                      className="run-id-options-wrapper"
                    >
                      <MultiSelect
                        buttonText={[Function]}
                        data={
                          Array [
                            Object {
                              "label": 1,
                              "selected": true,
                              "value": 1,
                            },
                            Object {
                              "label": 2,
                              "selected": true,
                              "value": 2,
                            },
                            Object {
                              "label": 3,
                              "selected": true,
                              "value": 3,
                            },
                            Object {
                              "label": 4,
                              "selected": true,
                              "value": 4,
                            },
                          ]
                        }
                        enableHTML={true}
                        id="run_ids"
                        includeSelectAllOption={true}
                        maxHeight={300}
                        multiple={true}
                        onChange={[Function]}
                        onDeselectAll={[Function]}
                        onSelectAll={[Function]}
                        selectedClass="run-id-selected"
                      />
                    </div>
                  </Well>
                  <MetricsPlotView
                    metricsResponse={
                      Array [
                        Object {
                          "_id": "5a2b5a8c9c7a505a652f6127",
                          "name": "pretrain.train.loss",
                          "run_id": 1,
                          "steps": Array [
                            0,
                            1,
                            2,
                            3,
                            4,
                          ],
                          "timestamps": Array [
                            "2017-12-09T03:37:44.425Z",
                            "2017-12-09T03:41:54.414Z",
                            "2017-12-09T03:46:01.766Z",
                            "2017-12-09T03:50:07.365Z",
                            "2017-12-09T03:54:12.560Z",
                          ],
                          "values": Array [
                            0.7159541824544438,
                            0.3840367944955761,
                            0.3469185283233073,
                            0.30483262065173106,
                            0.28915774130337507,
                          ],
                        },
                        Object {
                          "_id": "5a2b5aa09c7a505a652f6146",
                          "name": "pretrain.val.loss",
                          "run_id": 2,
                          "steps": Array [
                            0,
                            1,
                            2,
                            3,
                            4,
                          ],
                          "timestamps": Array [
                            "2017-12-09T03:38:01.945Z",
                            "2017-12-09T03:42:11.673Z",
                            "2017-12-09T03:46:18.843Z",
                            "2017-12-09T03:50:24.377Z",
                            "2017-12-09T03:54:29.752Z",
                          ],
                          "values": Array [
                            0.32177006650114165,
                            0.23237958704995795,
                            0.23340759051386187,
                            0.21925230575196739,
                            0.20541178824900605,
                          ],
                        },
                        Object {
                          "_id": "5a2b5a8c9c7a505a652f6127",
                          "name": "pretrain.train.loss",
                          "run_id": 3,
                          "steps": Array [
                            0,
                            1,
                            2,
                            3,
                            4,
                          ],
                          "timestamps": Array [
                            "2017-12-09T03:37:44.425Z",
                            "2017-12-09T03:41:54.414Z",
                            "2017-12-09T03:46:01.766Z",
                            "2017-12-09T03:50:07.365Z",
                            "2017-12-09T03:54:12.560Z",
                          ],
                          "values": Array [
                            0.7859541824544438,
                            0.5640367944955761,
                            0.8769185283233073,
                            0.604832620651731,
                            0.5891577413033751,
                          ],
                        },
                        Object {
                          "_id": "5a2b5aa09c7a505a652f6146",
                          "name": "pretrain.val.loss",
                          "run_id": 4,
                          "steps": Array [
                            0,
                            1,
                            2,
                            3,
                            4,
                          ],
                          "timestamps": Array [
                            "2017-12-09T03:38:01.945Z",
                            "2017-12-09T03:42:11.673Z",
                            "2017-12-09T03:46:18.843Z",
                            "2017-12-09T03:50:24.377Z",
                            "2017-12-09T03:54:29.752Z",
                          ],
                          "values": Array [
                            0.5217700665011417,
                            0.9237958704995795,
                            0.11340759051386187,
                            0.41925230575196737,
                            0.905411788249006,
                          ],
                        },
                      ]
                    }
                    runId=""
                    stateFilterKeys={
                      Array [
                        "selectedMetricNames",
                        "selectedXAxis",
                        "selectedYAxis",
                        "plotWidth",
                        "plotHeight",
                        "smoothing",
                      ]
                    }
                  />
                </div>
              </ProgressWrapper>
            </div>
          </div>
        </div>
      </div>
    </ProgressWrapper>
  </ModalBody>
</Modal>
`;
